src/Features/Core/Portable/FileHeaders/AbstractFileHeaderCodeFixProvider.cs

```diff
@@ -0,0 +1,243 @@
+                return root;
+            }
+
+            var expectedFileHeader = fileHeaderTemplate.Replace("{fileName}", Path.GetFileName(document.FilePath));
+
+            var fileHeader = FileHeaderHelper.ParseFileHeader(root);
+            SyntaxNode newSyntaxRoot;
+            if (fileHeader.IsMissing)
+            {
+                newSyntaxRoot = AddHeader(document, root, expectedFileHeader);
+            }
+            else
+            {
+                newSyntaxRoot = ReplaceHeader(document, root, expectedFileHeader);
+            }
+
+            return newSyntaxRoot;
+        }
+
+        private SyntaxNode ReplaceHeader(Document document, SyntaxNode root, string expectedFileHeader)
+        {
+            // Skip single line comments, whitespace, and end of line trivia until a blank line is encountered.
+            var trivia = root.GetLeadingTrivia();
+            var onBlankLine = false;
+            var inCopyright = true;
+            var removalList = new List<int>();
+            var leadingSpaces = string.Empty;
+            var possibleLeadingSpaces = string.Empty;
+
+            // Need to do this with index so we get the line endings correct.
+            for (var i = 0; i < trivia.Count; i++)
+            {
+                var triviaLine = trivia[i];
+                if (triviaLine.RawKind == SyntaxKinds.SingleLineCommentTrivia)
+                {
+                    if (possibleLeadingSpaces != string.Empty)
+                    {
+                        leadingSpaces = possibleLeadingSpaces;
+                    }
+
+                    removalList.Add(i);
+                    onBlankLine = false;
+                }
+                else if (triviaLine.RawKind == SyntaxKinds.WhitespaceTrivia)
+                {
+                    if (leadingSpaces == string.Empty)
+                    {
+                        possibleLeadingSpaces = triviaLine.ToFullString();
+                    }
+
+                    if (inCopyright)
+                    {
+                        removalList.Add(i);
+                    }
+                }
+                else if (triviaLine.RawKind == SyntaxKinds.EndOfLineTrivia)
+                {
+                    possibleLeadingSpaces = string.Empty;
+
+                    if (inCopyright)
+                    {
+                        removalList.Add(i);
+                    }
+
+                    if (onBlankLine)
+                    {
+                        break;
+                    }
+                    else
+                    {
+                        onBlankLine = true;
+                    }
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            // Remove copyright lines in reverse order.
+            for (var i = removalList.Count - 1; i >= 0; i--)
+            {
+                trivia = trivia.RemoveAt(removalList[i]);
```

Comment:
technically n^2.  but can't imagine that mattering.   consider though switching to just building the new list O(n) and just making a SyntaxTriviaList out of that.

Reply:
it's not perfect but I feel OK about leaving this one for now.

Reply:
fine with me :)